python/utils/train_model.py

```diff
@@ -0,0 +1,313 @@
+valid_dataloader = DataLoader(valid_data, sampler=valid_sampler, batch_size=BATCH_SIZE)
+
+print('Loaded training and validation data into DataLoaders successfully.')
+
+###
+
+# Initialize model
+model = BertForTokenClassification.from_pretrained('bert-base-cased',
+                                                   num_labels=NUM_LABELS)
+
+model.to(device)
+
+print(f'Initialized model and moved it to {device}.')
+
+###
+
+# Set hyperparameters (optimizer, weight decay, learning rate)
+
+if FULL_FINETUNING:
+    param_optimizer = list(model.named_parameters())
+    no_decay = ['bias', 'gamma', 'beta']
+    optimizer_grouped_parameters = [
+        {'params': [p for n, p in param_optimizer
+                    if not any(nd in n for nd in no_decay)],
+         'weight_decay_rate': 0.01},
+        {'params': [p for n, p in param_optimizer
+                    if any(nd in n for nd in no_decay)],
+         'weight_decay_rate': 0.0}
+    ]
+else:
+    param_optimizer = list(model.classifier.named_parameters())
+    optimizer_grouped_parameters = [{"params": [p for n, p in param_optimizer]}]
+
+optimizer = Adam(optimizer_grouped_parameters, lr=3e-5)
+
+print('Initialized optimizer and set hyperparameters.')
+
+###
+
+for _ in trange(EPOCHS, desc="Epoch"):
+
+    # Training loop
+    print('Starting training loop.')
+    model.train()
+    tr_loss = 0
+    nb_tr_examples, nb_tr_steps = 0, 0
+
+    for step, batch in enumerate(train_dataloader):
+
+        print('Starting a new training batch.')
+
+        # add batch to gpu
+        batch = tuple(t.to(device) for t in batch)
+        b_input_ids, b_input_mask, b_labels = batch
+
+        # forward pass
+        loss = model(b_input_ids, token_type_ids=None,
+                     attention_mask=b_input_mask, labels=b_labels)
+        # backward pass
+        loss.backward()
+
+        # track train loss
+        tr_loss += loss.item()
+        nb_tr_examples += b_input_ids.size(0)
+        nb_tr_steps += 1
+
+        # gradient clipping
+        torch.nn.utils.clip_grad_norm_(parameters=model.parameters(), max_norm=MAX_GRAD_NORM)
+
+        # update parameters
+        optimizer.step()
+        model.zero_grad()
+
+    # print train loss per epoch
+    print(f"Train loss: {tr_loss / nb_tr_steps}")
+
+
+    print('Starting validation loop.')
+
+    # VALIDATION on validation set
+    model.eval()
+    eval_loss, eval_accuracy = 0, 0
+    nb_eval_steps, nb_eval_examples = 0, 0
+    predictions, true_labels = [], []
+
+    for batch in valid_dataloader:
+
+        batch = tuple(t.to(device) for t in batch)
+        b_input_ids, b_input_mask, b_labels = batch
+
+        with torch.no_grad():
+            tmp_eval_loss = model(b_input_ids, token_type_ids=None,
+                                  attention_mask=b_input_mask, labels=b_labels)
+            logits = model(b_input_ids, token_type_ids=None,
+                           attention_mask=b_input_mask)
+
+        logits = logits.detach().cpu().numpy()
```

Comment:
error here: Train loss: 0.1815053676309194
Starting validation loop.
Traceback (most recent call last):
 File "train_model.py", line 281, in <module>
   logits = logits.detach().cpu().numpy()
AttributeError: 'tuple' object has no attribute 'detach'